nuclio/config.py

```diff
@@ -203,37 +203,47 @@ def set_env(config, env):
             raise ValueError(
                 'cannot parse environment value from: {}'.format(line))
 
+        # TODO: allow external source env with magic
         update_env_var(config, key, value)
 
 
 def set_env_dict(config, env={}):
     for k, v in env.items():
-        update_env_var(config, k, str(v))
+        update_env_var(config, k, value=str(v))
 
 
-def update_env_var(config, key, value):
-    i = 0
-    found = False
-    for v in config['spec']['env']:
-        if v['name'] == key:
-            found = True
-            break
-        i += 1
+def set_external_source_env_dict(config, external_source_env={}):
+    for k, v in external_source_env.items():
+        update_env_var(config, k, value_from=v)
 
-    item = {'name': key, 'value': value}
-    if found:
-        config['spec']['env'][i] = item
+
+def update_env_var(config, key, value=None, value_from=None):
+    if value:
+        item = {'name': key, 'value': value}
+    elif value_from:
+        item = {'name': key, 'valueFrom': value_from}
+    else:
+        message = 'env var requires either value or value_from'
+        logger.warning(message, config=config, key=key)
+        raise Exception(message)
+
+    # find key existence in env
+    location = next((idx for idx, env_var in enumerate(config['spec']['env']) if env_var['name'] == key), None)
+    if location is not None:
+        config['spec']['env'][location] = item
     else:
         config['spec']['env'].append(item)
 
 
-def fill_config(config, extra_config={}, env={}, cmd=[], mount: Volume = None):
+def fill_config(config, extra_config={}, env={}, cmd=[], mount: Volume = None, external_source_env={}):
```

Comment:
for the future - note that thought the original author of nuclio-jupyter decided to use mutable function arg objects it is highly not recommended.